src/cuda/cuda_sampling.cu

```diff
@@ -537,32 +540,534 @@ void LaunchSort(SamplingData* data, cudaStream_t stream, float* scores_in, float
+            if (scaled_score > max_val) max_val = scaled_score;
+        }
+
+        float sum_exp = 0.0f;
+        for (int i = 0; i < k; i++) {
+            sum_exp += expf(top_k_scores_reg[i] - max_val);
+        }
+
+        int final_offset = batch_idx * k;
+        for (int i = 0; i < k; i++) {
+            final_scores[final_offset + i] = expf(top_k_scores_reg[i] - max_val) / sum_exp;
+            final_indices[final_offset + i] = final_top_k.key[i];
+        }
+    }
+}
+
+// Single kernel fusion of Map and Reduce stages
+template <int max_k, int kBlockSize>
+__global__ void SingleKernelMapReduceTopKAndSoftmax(
+    int* final_indices,
+    float* final_scores,
+    const float* scores_in,
+    int* intermediate_indices,
+    float* intermediate_scores,
+    int vocab_size,
+    int num_partitions,
+    int k,
+    float temperature,
+    int* sync_counter) {
+    // --- STAGE 1: MAP (All blocks execute this) ---
+    const int batch_idx = 0; // Specialized for batch_size = 1
+    const int partition_idx = blockIdx.x;
+    const int num_intermediate_results_per_batch = num_partitions * max_k;
+
+    const int partition_size = (vocab_size + num_partitions - 1) / num_partitions;
+    const int start_index = partition_idx * partition_size;
+    const int end_index = min(start_index + partition_size, vocab_size);
+
+    TopK<float, max_k> thread_top_k;
+    thread_top_k.Init();
+
+    for (int i = start_index + threadIdx.x; i < end_index; i += kBlockSize) {
+        thread_top_k.Insert(scores_in[i], i);
+    }
+
+    typedef cub::BlockReduce<TopK<float, max_k>, kBlockSize> BlockReduce;
+    __shared__ typename BlockReduce::TempStorage temp_storage;
+    TopK<float, max_k> block_top_k = BlockReduce(temp_storage).Reduce(thread_top_k, reduce_topk_op<float, max_k>);
+
+    if (threadIdx.x == 0) {
+        int offset = partition_idx * max_k;
+        for (int i = 0; i < max_k; ++i) {
+            intermediate_scores[offset + i] = block_top_k.value[i];
+            intermediate_indices[offset + i] = block_top_k.key[i];
+        }
+    }
+    __syncthreads();
+
+    // --- STAGE 1.5: GLOBAL SYNC ---
+    if (threadIdx.x == 0) {
+        atomicAdd(sync_counter, 1);
+    }
+
+    // --- STAGE 2: REDUCE (Only block 0 executes this) ---
+    if (blockIdx.x == 0) {
+        // Block 0 waits for all other blocks to report in
+        volatile int* p_counter = sync_counter;
+        while (*p_counter < gridDim.x) {}
+        __threadfence(); // Ensure visibility of writes from other blocks
+
+        // Now perform the reduction
+        TopK<float, max_k> final_thread_top_k;
+        final_thread_top_k.Init();
+
+        for (int i = threadIdx.x; i < num_intermediate_results_per_batch; i += kBlockSize) {
+            final_thread_top_k.Insert(intermediate_scores[i], intermediate_indices[i]);
+        }
+
+        TopK<float, max_k> final_top_k = BlockReduce(temp_storage).Reduce(final_thread_top_k, reduce_topk_op<float, max_k>);
```

Comment:
Suggestion for one more variant of this: If k is small enough and the number of partitions are small enough - I think a single thread doing the reduction may work better than again doing multi-stage reduction and incurring that overhead. So, maybe having a variant to do single thread reduction will help ?